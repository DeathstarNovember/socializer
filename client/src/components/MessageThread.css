.chat-messages {
  flex: 1;
  overflow-y: scroll;
}

.chat-self {
  flex-direction: column;
  align-items: flex-end;
}

.chat-self .chat-bubble-wrapper {
  flex-direction: row-reverse;
}

.chat-self .chat-bubble {
  margin-right: 0.25rem;
  background-color: #007bff;
  color: white;
}

.chat-others {
  align-items: flex-start;
}

.chat-others .chat-bubble {
  margin-left: 0.25rem;
  background-color: #f1f0f0;
}

.chat-bubble {
  border-radius: 5px;
}

.avatar-spacer {
  width: 30px;
}
