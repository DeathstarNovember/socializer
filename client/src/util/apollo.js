import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
