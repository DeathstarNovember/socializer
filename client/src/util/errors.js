import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
