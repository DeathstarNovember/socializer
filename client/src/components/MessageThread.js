import React, { useContext, useLayoutEffect, useRef } from "react";
import cx from "classnames";
import { OverlayTrigger, Tooltip } from "react-bootstrap";
import Gravatar from "react-gravatar";
import { AuthContext } from "util/context";
import "./MessageThread.css";

const MessageThread = ({ messages }) => {
  const messageThread = useRef();
  const { userId } = useContext(AuthContext);

  useLayoutEffect(() => {
    messageThread.current.scrollTop = messageThread.current.scrollHeight;
  });

  return (
    <div ref={messageThread} className="chat-messages">
      {messages.map((message, idx) => (
        <div
          key={message.id}
          className={cx("d-flex", {
            "chat-self": message.user.id === String(userId),
            "chat-others": message.user.id !== String(userId),
          })}
        >
          <div className="d-flex mb-1 chat-bubble-wrapper">
            {idx === 0 || messages[idx - 1].user.id !== message.user.id ? (
              <OverlayTrigger
                placement="bottom"
                overlay={
                  <Tooltip id={`user-${message.user.id}`}>
                    {message.user.name}
                  </Tooltip>
                }
              >
                <Gravatar
                  md5={message.user.gravatarMd5}
                  className="rounded-circle mt-1"
                  size={30}
                />
              </OverlayTrigger>
            ) : (
              <div className="avatar-spacer" />
            )}
            <div className="p-2 chat-bubble">{message.body}</div>
          </div>
        </div>
      ))}
    </div>
  );
};

export default MessageThread;
