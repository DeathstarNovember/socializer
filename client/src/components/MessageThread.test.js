import React from "react";
import { render } from "@testing-library/react";
import { AuthContext } from "util/context";
import MessageThread from "./MessageThread";

describe("MessageThread", () => {
  const messages = [
    {
      id: 123,
      body: "Some thoughts",
      user: {
        id: "456",
        name: "John Smith",
        gravatarMd5: "abcdefg",
      },
    },
    {
      id: 456,
      body: "Other thoughts",
      user: {
        id: "789",
        name: "Jane Smith",
        gravatarMd5: "defghi",
      },
    },
  ];

  it("renders correctly", () => {
    const { container } = render(
      <AuthContext.Provider value={{ userId: "456" }}>
        <MessageThread messages={messages} />
      </AuthContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
