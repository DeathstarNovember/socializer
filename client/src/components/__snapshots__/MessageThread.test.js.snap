// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageThread renders correctly 1`] = `
<div>
  <div
    class="chat-messages"
  >
    <div
      class="d-flex chat-self"
    >
      <div
        class="d-flex mb-1 chat-bubble-wrapper"
      >
        <img
          alt="Gravatar for undefined"
          class="react-gravatar rounded-circle mt-1"
          height="30"
          src="//www.gravatar.com/avatar/abcdefg?d=retro&r=g&s=30"
          srcset="//www.gravatar.com/avatar/abcdefg?d=retro&r=g&s=60 2x"
          width="30"
        />
        <div
          class="p-2 chat-bubble"
        >
          Some thoughts
        </div>
      </div>
    </div>
    <div
      class="d-flex chat-others"
    >
      <div
        class="d-flex mb-1 chat-bubble-wrapper"
      >
        <img
          alt="Gravatar for undefined"
          class="react-gravatar rounded-circle mt-1"
          height="30"
          src="//www.gravatar.com/avatar/defghi?d=retro&r=g&s=30"
          srcset="//www.gravatar.com/avatar/defghi?d=retro&r=g&s=60 2x"
          width="30"
        />
        <div
          class="p-2 chat-bubble"
        >
          Other thoughts
        </div>
      </div>
    </div>
  </div>
</div>
`;
